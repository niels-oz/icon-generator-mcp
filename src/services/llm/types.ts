import { LLMResponse } from '../../types';

export interface LLM {
  generate(prompt: string, svgReferences?: string[], styleName?: string): Promise<LLMResponse>;
}
