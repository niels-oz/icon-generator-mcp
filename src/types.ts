// Core types for the icon generator MCP server with Sequential Thinking architecture

export interface IconGenerationRequest {
  png_paths: string[];
  prompt: string;
  output_name?: string;
  output_path?: string;
  style?: string;
}

export interface IconGenerationResponse {
  success: boolean;
  output_path?: string;
  message: string;
  processing_time?: number;
  error?: string;
  steps?: GenerationStep[];
}

export interface ConversionResult {
  svg: string;
  filename: string;
}

export interface LLMResponse {
  svg: string;
  filename: string;
}

// Sequential Thinking building blocks for icon generation
export interface GenerationStep {
  step: GenerationPhase;
  status: StepStatus;
  message: string;
  timestamp: Date;
  details?: any;
}

export type GenerationPhase = 
  | 'validation'
  | 'analysis'
  | 'conversion'
  | 'generation'
  | 'refinement'
  | 'output';

export type StepStatus = 'pending' | 'in_progress' | 'completed' | 'failed' | 'skipped';

export interface GenerationState {
  sessionId: string;
  request: IconGenerationRequest;
  steps: GenerationStep[];
  currentPhase: GenerationPhase;
  startTime: Date;
  context: GenerationContext;
}

export interface GenerationContext {
  validatedFiles: string[];
  svgReferences: string[];
  generatedSvg?: string;
  suggestedFilename?: string;
  outputPath?: string;
  errors: string[];
}