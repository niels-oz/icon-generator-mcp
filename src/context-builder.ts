import { StyleConfig, getStyleConfig } from './styles/few-shot-examples';

export interface GenerationContext {
  prompt: string;
  instructions: string;
  processing_info: {
    references_processed: number;
    style_applied: string | null;
    analysis_time_ms: number;
  };
}

/**
 * Build context for LLM generation based on user prompt, references, and style
 * This replaces the CLI-based LLM integration with context preparation
 */
export function buildGenerationContext(
  userPrompt: string, 
  svgReferences: string[] = [], 
  styleName?: string
): GenerationContext {
  const startTime = Date.now();
  
  // Get style config if specified
  const styleConfig = styleName ? getStyleConfig(styleName) : undefined;

  let systemPrompt = `You are an expert SVG icon designer. Given SVG references and a user request, generate a clean, optimized SVG icon.`;

  // Add few-shot examples if style config is provided
  if (styleConfig) {
    systemPrompt += `\n\nSTYLE: ${styleConfig.name}
${styleConfig.description}

Here are examples of this style:

${styleConfig.examples.map((example: any, index: number) => `Example ${index + 1}:
Prompt: "${example.prompt}"
Description: ${example.description}
SVG:
${example.svg}
`).join('\n')}

Follow the exact same style, structure, and visual approach as these examples.`;
  }

  // Add reference SVGs if provided
  if (svgReferences.length > 0) {
    systemPrompt += `\n\nReference SVG icons:
${svgReferences.map((svg, index) => `Reference ${index + 1}:
${svg}
`).join('\n')}`;
  }

  systemPrompt += `\n\nUser request: ${userPrompt}

Please generate:
1. A clean SVG icon that fulfills the request
2. A descriptive filename (no extension)

Format your response exactly as:
FILENAME: [suggested-filename]
SVG: [complete SVG code]

Requirements:
- Use proper SVG namespace: xmlns="http://www.w3.org/2000/svg"
- Include viewBox attribute for scalability
- Use clean, optimized SVG code
- No script tags or dangerous elements
- Filename should be descriptive and kebab-case${styleConfig ? `
- CRITICALLY IMPORTANT: Follow the exact style, stroke-width, color scheme, and visual approach from the provided examples` : ''}`;

  const analysisTime = Date.now() - startTime;

  return {
    prompt: systemPrompt,
    instructions: 'Generate an SVG icon based on the above context and return it in the specified format.',
    processing_info: {
      references_processed: svgReferences.length,
      style_applied: styleName || null,
      analysis_time_ms: analysisTime
    }
  };
}