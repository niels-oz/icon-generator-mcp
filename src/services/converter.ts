import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { spawn, execSync } from 'child_process';
import Jimp from 'jimp';

export class ConversionService {
  private readonly MAX_FILE_SIZE = 10 * 1024 * 1024; // 10MB
  
  async convertPNGToSVG(pngPath: string): Promise<string> {
    // Validate file exists
    if (!fs.existsSync(pngPath)) {
      throw new Error(`File not found: ${pngPath}`);
    }

    // Check file size
    const stats = fs.statSync(pngPath);
    if (stats.size > this.MAX_FILE_SIZE) {
      throw new Error(`File too large: ${pngPath}. Maximum size is ${this.MAX_FILE_SIZE / (1024 * 1024)}MB`);
    }

    // Validate PNG format using Jimp
    try {
      const image = await Jimp.read(pngPath);
      if (!image) {
        throw new Error('Invalid PNG format');
      }
    } catch (error) {
      throw new Error(`Invalid PNG format: ${pngPath}`);
    }

    // Find potrace binary
    const potracePath = this.findPotraceBinary();
    
    // Create temporary BMP file using cross-platform temp directory
    const tempBmpPath = path.join(os.tmpdir(), `temp-${Date.now()}.bmp`);
    
    try {
      // Convert PNG to BMP using Jimp
      const image = await Jimp.read(pngPath);
      await image.writeAsync(tempBmpPath);
      
      // Execute potrace to convert BMP to SVG with enhanced error handling
      const svgOutput = await this.executePotrace(potracePath, tempBmpPath);
      
      // Clean up temporary file
      if (fs.existsSync(tempBmpPath)) {
        fs.unlinkSync(tempBmpPath);
      }
      
      return svgOutput;
    } catch (error) {
      // Clean up temporary file on error
      if (fs.existsSync(tempBmpPath)) {
        fs.unlinkSync(tempBmpPath);
      }
      throw new Error(`Potrace conversion failed: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  findPotraceBinary(): string {
    try {
      const potracePath = execSync('which potrace', {
        encoding: 'utf8',
        stdio: ['pipe', 'pipe', 'pipe']
      }).trim();
      if (!potracePath) {
        throw new Error('Potrace not found');
      }
      return potracePath;
    } catch (error) {
      const err = error as any;
      if (err.code === 'ENOENT') {
        throw new Error('Potrace not found in PATH. Please install with: brew install potrace');
      }
      throw new Error(`Failed to locate potrace binary: ${err.message || 'Unknown error'}`);
    }
  }

  /**
   * Execute potrace command with enhanced error handling
   */
  private async executePotrace(potracePath: string, bmpPath: string): Promise<string> {
    return new Promise((resolve, reject) => {
      const args = ['--svg', '--output=-', bmpPath];
      const process = spawn(potracePath, args, {
        stdio: ['pipe', 'pipe', 'pipe']
      });

      let stdout = '';
      let stderr = '';

      process.stdout.on('data', (data) => {
        stdout += data.toString();
      });

      process.stderr.on('data', (data) => {
        stderr += data.toString();
      });

      process.on('close', (code) => {
        if (code === 0) {
          if (!stdout.trim()) {
            reject(new Error('Potrace produced no output'));
            return;
          }
          resolve(stdout);
        } else {
          const errorMessage = stderr.trim() || `Potrace exited with code ${code}`;
          reject(new Error(`Potrace conversion failed: ${errorMessage}`));
        }
      });

      process.on('error', (error) => {
        reject(new Error(`Failed to execute potrace: ${error.message}`));
      });

      // Set timeout for the process
      const timeout = setTimeout(() => {
        process.kill('SIGTERM');
        reject(new Error('Potrace conversion timed out after 30 seconds'));
      }, 30000);

      process.on('close', () => {
        clearTimeout(timeout);
      });
    });
  }
}