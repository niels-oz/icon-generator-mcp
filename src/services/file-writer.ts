import * as fs from 'fs';
import * as path from 'path';

export interface SaveResult {
  success: boolean;
  outputPath?: string;
  error?: string;
}

export class FileWriterService {
  /**
   * Sanitize filename to remove invalid filesystem characters
   */
  sanitizeFilename(filename: string): string {
    // Replace invalid characters with dash
    let sanitized = filename.replace(/[<>:"|*?/\\'/]/g, '-');
    
    // Replace multiple consecutive dashes with single dash
    sanitized = sanitized.replace(/-+/g, '-');
    
    // Replace spaces with dashes
    sanitized = sanitized.replace(/\s+/g, '-');
    
    // Remove leading/trailing dashes
    sanitized = sanitized.replace(/^-+|-+$/g, '');
    
    // If filename is empty or only invalid chars, use fallback
    if (!sanitized || sanitized.length === 0) {
      sanitized = 'generated-icon';
    }
    
    return sanitized;
  }

  /**
   * Generate output path based on filename and reference location
   */
  generateOutputPath(filename: string, referenceLocation: string | undefined | null): string {
    const sanitizedFilename = this.sanitizeFilename(filename);
    
    // If no reference location provided, use current directory
    if (!referenceLocation) {
      const finalFilename = sanitizedFilename.endsWith('.svg') 
        ? sanitizedFilename 
        : `${sanitizedFilename}.svg`;
      return `./${finalFilename}`;
    }
    
    // If reference location is a directory, use it directly
    // If it's a file, extract the directory
    const outputDir = path.extname(referenceLocation) 
      ? path.dirname(referenceLocation) 
      : referenceLocation;
    
    // Ensure .svg extension
    const finalFilename = sanitizedFilename.endsWith('.svg') 
      ? sanitizedFilename 
      : `${sanitizedFilename}.svg`;
    
    return path.join(outputDir, finalFilename);
  }

  /**
   * Resolve filename conflicts by appending index numbers
   */
  resolveNameConflicts(filePath: string): string {
    if (!fs.existsSync(filePath)) {
      return filePath;
    }

    const dir = path.dirname(filePath);
    const ext = path.extname(filePath);
    const baseName = path.basename(filePath, ext);

    let counter = 2;
    let resolvedPath: string;

    do {
      resolvedPath = path.join(dir, `${baseName}-${counter}${ext}`);
      counter++;
    } while (fs.existsSync(resolvedPath));

    return resolvedPath;
  }

  /**
   * Validate that reference path exists
   */
  validateReferencePath(referencePath: string): boolean {
    return fs.existsSync(referencePath);
  }

  /**
   * Extract directory from reference path
   */
  extractDirectory(referencePath: string): string {
    return path.dirname(referencePath);
  }

  /**
   * Check if directory is writable
   */
  private checkDirectoryWritable(dirPath: string): void {
    try {
      fs.accessSync(dirPath, fs.constants.W_OK);
    } catch (error) {
      throw new Error(`Cannot write to directory: ${dirPath}`);
    }
  }

  /**
   * Write SVG content to file
   */
  async saveGeneratedSVG(outputPath: string, svgContent: string): Promise<void> {
    const dir = path.dirname(outputPath);
    
    // Check directory write permissions
    this.checkDirectoryWritable(dir);
    
    try {
      fs.writeFileSync(outputPath, svgContent, 'utf8');
    } catch (error) {
      throw new Error(`Failed to write SVG file: ${error instanceof Error ? error.message : 'Unknown error'}`);
    }
  }

  /**
   * Complete workflow: generate path, resolve conflicts, and save file
   */
  async saveGeneratedIcon(
    suggestedFilename: string,
    referenceLocation: string | undefined,
    svgContent: string
  ): Promise<SaveResult> {
    try {
      // Validate reference location exists (if it's a file path)
      if (referenceLocation && path.extname(referenceLocation) && !this.validateReferencePath(referenceLocation)) {
        return {
          success: false,
          error: `Reference file not found: ${referenceLocation}`
        };
      }

      // Generate output path
      const outputPath = this.generateOutputPath(suggestedFilename, referenceLocation);
      
      // Resolve any naming conflicts
      const resolvedPath = this.resolveNameConflicts(outputPath);
      
      // Save the file
      await this.saveGeneratedSVG(resolvedPath, svgContent);
      
      return {
        success: true,
        outputPath: resolvedPath
      };
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error during file save'
      };
    }
  }
}