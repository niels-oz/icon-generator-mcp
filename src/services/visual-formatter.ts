import chalk from 'chalk';
import { GenerationStep, GenerationPhase, StepStatus, GenerationState } from '../types';

export class VisualFormatter {
  private static readonly PHASE_ICONS = {
    validation: '🔍',
    analysis: '📊',
    generation: '🎨',
    refinement: '✨',
    output: '💾'
  };

  private static readonly PHASE_COLORS = {
    validation: chalk.blue,
    analysis: chalk.cyan,
    generation: chalk.magenta,
    refinement: chalk.green,
    output: chalk.blue
  };

  private static readonly STATUS_ICONS = {
    pending: '⏳',
    in_progress: '🔄',
    completed: '✅',
    failed: '❌',
    skipped: '⏭️'
  };

  private static readonly STATUS_COLORS = {
    pending: chalk.gray,
    in_progress: chalk.yellow,
    completed: chalk.green,
    failed: chalk.red,
    skipped: chalk.gray
  };

  formatStep(step: GenerationStep): string {
    const phaseIcon = VisualFormatter.PHASE_ICONS[step.step];
    const statusIcon = VisualFormatter.STATUS_ICONS[step.status];
    const phaseColor = VisualFormatter.PHASE_COLORS[step.step];
    const statusColor = VisualFormatter.STATUS_COLORS[step.status];

    const phaseText = phaseColor(step.step.toUpperCase());
    const statusText = statusColor(step.status.replace('_', ' ').toUpperCase());
    const timestamp = chalk.gray(step.timestamp.toLocaleTimeString());

    return `${phaseIcon} ${statusIcon} ${phaseText} - ${statusText} ${timestamp}\n   ${chalk.white(step.message)}`;
  }

  formatProgress(state: GenerationState): string {
    const totalSteps = 5; // Number of phases (validation, analysis, generation, refinement, output)
    const completedSteps = state.steps.filter(s => s.status === 'completed').length;
    const progress = Math.round((completedSteps / totalSteps) * 100);
    
    const progressBar = this.createProgressBar(progress);
    const currentPhase = VisualFormatter.PHASE_COLORS[state.currentPhase](state.currentPhase.toUpperCase());

    return `
${chalk.bold.blue('🎯 Icon Generation Progress')}
${progressBar} ${chalk.bold(`${progress}%`)}
${chalk.gray(`Current Phase: ${currentPhase}`)}
${chalk.gray(`Session: ${state.sessionId}`)}
${chalk.gray(`Elapsed: ${this.formatDuration(Date.now() - state.startTime.getTime())}`)}
    `.trim();
  }

  formatGenerationSummary(state: GenerationState): string {
    const header = chalk.bold.blue('📋 Generation Summary');
    const stepsFormatted = state.steps.map(step => this.formatStep(step)).join('\n\n');
    
    let summary = `${header}\n${chalk.gray(`Session: ${state.sessionId}`)}\n\n${stepsFormatted}`;

    // Add context information if available
    if (state.context.validatedFiles.length > 0) {
      summary += `\n\n${chalk.bold('📁 Input Files:')}\n${state.context.validatedFiles.map(f => `   • ${f}`).join('\n')}`;
    }

    if (state.context.errors.length > 0) {
      summary += `\n\n${chalk.bold.red('⚠️ Errors:')}\n${state.context.errors.map(e => `   • ${e}`).join('\n')}`;
    }

    return summary;
  }

  formatPhaseTransition(fromPhase: GenerationPhase, toPhase: GenerationPhase): string {
    const fromIcon = VisualFormatter.PHASE_ICONS[fromPhase];
    const toIcon = VisualFormatter.PHASE_ICONS[toPhase];
    const fromColor = VisualFormatter.PHASE_COLORS[fromPhase];
    const toColor = VisualFormatter.PHASE_COLORS[toPhase];

    return `${fromIcon} ${fromColor(fromPhase)} ${chalk.gray('→')} ${toIcon} ${toColor(toPhase)}`;
  }

  formatFinalResult(state: GenerationState, success: boolean): string {
    const icon = success ? '🎉' : '💥';
    const status = success ? chalk.green.bold('SUCCESS') : chalk.red.bold('FAILED');
    const duration = this.formatDuration(Date.now() - state.startTime.getTime());

    let result = `
${icon} ${status} ${icon}
${chalk.gray(`Duration: ${duration}`)}
    `.trim();

    if (success && state.context.outputPath) {
      result += `\n${chalk.bold.green('📁 Saved to:')} ${chalk.underline(state.context.outputPath)}`;
    }

    if (!success && state.context.errors.length > 0) {
      result += `\n${chalk.bold.red('❌ Final Error:')} ${state.context.errors[state.context.errors.length - 1]}`;
    }

    return result;
  }

  private createProgressBar(percentage: number, width = 20): string {
    const filled = Math.round((percentage / 100) * width);
    const empty = width - filled;
    
    return chalk.green('█'.repeat(filled)) + chalk.gray('░'.repeat(empty));
  }

  private formatDuration(milliseconds: number): string {
    const seconds = Math.floor(milliseconds / 1000);
    const minutes = Math.floor(seconds / 60);
    
    if (minutes > 0) {
      return `${minutes}m ${seconds % 60}s`;
    }
    return `${seconds}s`;
  }

  // Helper method for creating boxed output
  formatBox(title: string, content: string, color: typeof chalk.blue = chalk.blue): string {
    const lines = content.split('\n');
    const maxLength = Math.max(title.length, ...lines.map(line => this.stripAnsi(line).length));
    const width = Math.min(maxLength + 4, 80);
    
    const border = '─'.repeat(width - 2);
    const paddedTitle = ` ${title} `.padEnd(width - 2);
    
    let result = color(`┌${border}┐\n`);
    result += color(`│${paddedTitle}│\n`);
    result += color(`├${border}┤\n`);
    
    lines.forEach(line => {
      const paddedLine = ` ${line} `.padEnd(width - 2);
      result += color('│') + paddedLine.substring(0, width - 2) + color('│\n');
    });
    
    result += color(`└${border}┘`);
    
    return result;
  }

  private stripAnsi(text: string): string {
    // Simple ANSI escape code removal for length calculation
    return text.replace(/\u001b\[[0-9;]*m/g, '');
  }
}