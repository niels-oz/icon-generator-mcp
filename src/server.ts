import { IconGenerationRequest, IconGenerationResponse } from './types';
import { ConversionService } from './services/converter';
import { LLMService } from './services/llm';
import { FileWriterService } from './services/file-writer';
import * as path from 'path';

export class MCPServer {
  public readonly name = 'icon-generator-mcp';
  public readonly version = '0.1.0';
  
  private conversionService: ConversionService;
  private llmService: LLMService;
  private fileWriterService: FileWriterService;
  
  constructor() {
    this.conversionService = new ConversionService();
    this.llmService = new LLMService();
    this.fileWriterService = new FileWriterService();
  }

  getTools() {
    return [
      {
        name: 'generate_icon',
        description: 'Generate SVG icon from PNG references and text prompt',
        inputSchema: {
          type: 'object',
          properties: {
            png_paths: { 
              type: 'array', 
              items: { type: 'string' },
              description: 'Array of PNG file paths to use as references'
            },
            prompt: { 
              type: 'string',
              description: 'Text description of the desired icon'
            },
            output_name: { 
              type: 'string',
              description: 'Optional custom output filename'
            },
            output_path: { 
              type: 'string',
              description: 'Optional custom output directory path'
            }
          },
          required: ['png_paths', 'prompt']
        }
      }
    ];
  }

  async handleToolCall(toolName: string, request: any): Promise<IconGenerationResponse> {
    const startTime = Date.now();
    
    try {
      if (toolName !== 'generate_icon') {
        return {
          success: false,
          message: 'Tool not found',
          error: `Unknown tool: ${toolName}`
        };
      }

      const validatedRequest = this.validateRequest(request);
      const result = await this.generateIcon(validatedRequest);
      
      return {
        success: true,
        output_path: result.output_path,
        message: result.message,
        processing_time: Date.now() - startTime
      };
    } catch (error) {
      return {
        success: false,
        message: 'Icon generation failed',
        error: error instanceof Error ? error.message : 'Unknown error',
        processing_time: Date.now() - startTime
      };
    }
  }

  private validateRequest(request: any): IconGenerationRequest {
    if (!request.png_paths || !Array.isArray(request.png_paths) || request.png_paths.length === 0) {
      throw new Error('png_paths is required and must be a non-empty array');
    }

    if (!request.prompt || typeof request.prompt !== 'string' || request.prompt.trim() === '') {
      throw new Error('prompt is required and must be a non-empty string');
    }

    return {
      png_paths: request.png_paths,
      prompt: request.prompt.trim(),
      output_name: request.output_name,
      output_path: request.output_path
    };
  }

  private async generateIcon(request: IconGenerationRequest): Promise<{ output_path: string; message: string }> {
    // Step 1: Convert all PNG files to SVG references
    const svgReferences: string[] = [];
    for (const pngPath of request.png_paths) {
      try {
        const svgContent = await this.conversionService.convertPNGToSVG(pngPath);
        svgReferences.push(svgContent);
      } catch (error) {
        throw new Error(`Failed to convert PNG file ${pngPath}: ${error instanceof Error ? error.message : 'Unknown error'}`);
      }
    }

    // Step 2: Generate icon using LLM with SVG references and prompt
    const llmResponse = await this.llmService.generateSVG(request.prompt, svgReferences);

    // Step 3: Determine output filename
    const outputFilename = request.output_name || llmResponse.filename;

    // Step 4: Save generated SVG to file
    const referenceForLocation = request.output_path || request.png_paths[0];
    const saveResult = await this.fileWriterService.saveGeneratedIcon(
      outputFilename,
      referenceForLocation,
      llmResponse.svg
    );

    if (!saveResult.success) {
      throw new Error(saveResult.error || 'Failed to save generated icon');
    }

    // Step 5: Return success response
    return {
      output_path: saveResult.outputPath!,
      message: `Icon generated successfully: ${path.basename(saveResult.outputPath!)}`
    };
  }
}