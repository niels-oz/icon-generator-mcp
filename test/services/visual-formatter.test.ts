import { VisualFormatter } from '../../src/services/visual-formatter';
import { GenerationStep, GenerationState, GenerationPhase, StepStatus } from '../../src/types';
import chalk from 'chalk';

// Mock chalk to ensure consistent snapshot output
jest.mock('chalk', () => {
  const mockChalk = {
    blue: Object.assign(jest.fn((text) => `[BLUE]${text}[/BLUE]`), { bold: jest.fn((text) => `[BLUE_BOLD]${text}[/BLUE_BOLD]`) }),
    cyan: jest.fn((text) => `[CYAN]${text}[/CYAN]`),
    magenta: jest.fn((text) => `[MAGENTA]${text}[/MAGENTA]`),
    green: Object.assign(jest.fn((text) => `[GREEN]${text}[/GREEN]`), { bold: jest.fn((text) => `[GREEN_BOLD]${text}[/GREEN_BOLD]`) }),
    red: Object.assign(jest.fn((text) => `[RED]${text}[/RED]`), { bold: jest.fn((text) => `[RED_BOLD]${text}[/RED_BOLD]`) }),
    yellow: jest.fn((text) => `[YELLOW]${text}[/YELLOW]`),
    gray: jest.fn((text) => `[GRAY]${text}[/GRAY]`),
    white: jest.fn((text) => `[WHITE]${text}[/WHITE]`),
    underline: jest.fn((text) => `[UNDERLINE]${text}[/UNDERLINE]`),
    bold: Object.assign(
      jest.fn((text) => `[BOLD]${text}[/BOLD]`),
      {
        blue: jest.fn((text) => `[BOLD_BLUE]${text}[/BOLD_BLUE]`),
        green: jest.fn((text) => `[BOLD_GREEN]${text}[/BOLD_GREEN]`),
        red: jest.fn((text) => `[BOLD_RED]${text}[/BOLD_RED]`),
      }
    )
  };
  return mockChalk;
});

describe('Visual Formatter Service', () => {
  let formatter: VisualFormatter;
  let mockState: GenerationState;

  beforeEach(() => {
    // Mock Date.now() to return consistent timestamps for snapshots
    jest.spyOn(Date, 'now').mockReturnValue(new Date('2025-08-16T10:00:00Z').getTime());
    formatter = new VisualFormatter();
    
    mockState = {
      sessionId: 'test-session-123',
      request: {
        prompt: 'Create a star icon',
        reference_paths: ['/test/file1.png', '/test/file2.svg'],
        output_name: undefined,
        output_path: undefined,
        style: undefined
      },
      currentPhase: 'validation' as GenerationPhase,
      startTime: new Date('2025-08-16T09:59:00Z'),
      steps: [
        {
          step: 'validation' as GenerationPhase,
          status: 'completed' as StepStatus,
          message: 'Validated 2 input files and prompt',
          timestamp: new Date('2025-08-16T10:00:00Z'),
          details: undefined
        }
      ],
      context: {
        validatedFiles: ['/test/file1.png', '/test/file2.svg'],
        svgReferences: [],
        errors: [],
        generatedSvg: undefined,
        suggestedFilename: undefined,
        outputPath: undefined
      }
    };
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  describe('Core Formatting Behaviors', () => {
    it('should format a single step correctly', () => {
      const step = mockState.steps[0];
      const result = formatter.formatStep(step);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('🔍'); // validation icon
      expect(result).toContain('✅'); // completed icon
      expect(result).toContain('VALIDATION');
      expect(result).toContain('COMPLETED');
      expect(result).toContain('Validated 2 input files and prompt');
    });

    it('should format the overall progress display', () => {
      const stateWith3Completed = {
        ...mockState,
        steps: [
          { ...mockState.steps[0], step: 'validation' as GenerationPhase, status: 'completed' as StepStatus },
          { ...mockState.steps[0], step: 'analysis' as GenerationPhase, status: 'completed' as StepStatus },
          { ...mockState.steps[0], step: 'generation' as GenerationPhase, status: 'completed' as StepStatus }
        ]
      };

      const result = formatter.formatProgress(stateWith3Completed);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('🎯 Icon Generation Progress');
      expect(result).toContain('60%'); // 3/5 phases = 60%
      expect(result).toContain('Session: test-session-123');
    });

    it('should format the final generation summary for a successful run', () => {
      const result = formatter.formatGenerationSummary(mockState);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('📋 Generation Summary');
      expect(result).toContain('Session: test-session-123');
      expect(result).toContain('📁 Input Files:');
      expect(result).toContain('/test/file1.png');
      expect(result).toContain('/test/file2.svg');
    });

    it('should format the final generation summary for a failed run', () => {
      const failedState = {
        ...mockState,
        context: {
          ...mockState.context,
          errors: ['File not found', 'LLM generation failed']
        }
      };

      const result = formatter.formatGenerationSummary(failedState);
      
      expect(result).toMatchSnapshot();
      expect(result).toContain('⚠️ Errors:');
      expect(result).toContain('File not found');
      expect(result).toContain('LLM generation failed');
    });

    it('should format the final result message (success and failure)', () => {
      // Test successful result
      const successState = {
        ...mockState,
        context: { ...mockState.context, outputPath: '/test/output/star-icon.svg' }
      };

      const successResult = formatter.formatFinalResult(successState, true);
      expect(successResult).toMatchSnapshot();
      expect(successResult).toContain('🎉');
      expect(successResult).toContain('SUCCESS');
      expect(successResult).toContain('📁 Saved to:');

      // Test failed result
      const failedState = {
        ...mockState,
        context: { ...mockState.context, errors: ['Critical error'] }
      };

      const failedResult = formatter.formatFinalResult(failedState, false);
      expect(failedResult).toMatchSnapshot();
      expect(failedResult).toContain('💥');
      expect(failedResult).toContain('FAILED');
      expect(failedResult).toContain('❌ Final Error:');
      expect(failedResult).toContain('Critical error');
    });
  });
});