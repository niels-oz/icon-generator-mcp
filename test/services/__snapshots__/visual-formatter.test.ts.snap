// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visual Formatter Service Core Formatting Behaviors should format a single step correctly 1`] = `
"🔍 ✅ [BLUE]VALIDATION[/BLUE] - [GREEN]COMPLETED[/GREEN] [GRAY]12:00:00 PM[/GRAY]
   [WHITE]Validated 2 input files and prompt[/WHITE]"
`;

exports[`Visual Formatter Service Core Formatting Behaviors should format the final generation summary for a failed run 1`] = `
"[BOLD_BLUE]📋 Generation Summary[/BOLD_BLUE]
[GRAY]Session: test-session-123[/GRAY]

🔍 ✅ [BLUE]VALIDATION[/BLUE] - [GREEN]COMPLETED[/GREEN] [GRAY]12:00:00 PM[/GRAY]
   [WHITE]Validated 2 input files and prompt[/WHITE]

[BOLD]📁 Input Files:[/BOLD]
   • /test/file1.png
   • /test/file2.svg

[BOLD_RED]⚠️ Errors:[/BOLD_RED]
   • File not found
   • LLM generation failed"
`;

exports[`Visual Formatter Service Core Formatting Behaviors should format the final generation summary for a successful run 1`] = `
"[BOLD_BLUE]📋 Generation Summary[/BOLD_BLUE]
[GRAY]Session: test-session-123[/GRAY]

🔍 ✅ [BLUE]VALIDATION[/BLUE] - [GREEN]COMPLETED[/GREEN] [GRAY]12:00:00 PM[/GRAY]
   [WHITE]Validated 2 input files and prompt[/WHITE]

[BOLD]📁 Input Files:[/BOLD]
   • /test/file1.png
   • /test/file2.svg"
`;

exports[`Visual Formatter Service Core Formatting Behaviors should format the final result message (success and failure) 1`] = `
"🎉 [GREEN_BOLD]SUCCESS[/GREEN_BOLD] 🎉
[GRAY]Duration: 1m 0s[/GRAY]
[BOLD_GREEN]📁 Saved to:[/BOLD_GREEN] [UNDERLINE]/test/output/star-icon.svg[/UNDERLINE]"
`;

exports[`Visual Formatter Service Core Formatting Behaviors should format the final result message (success and failure) 2`] = `
"💥 [RED_BOLD]FAILED[/RED_BOLD] 💥
[GRAY]Duration: 1m 0s[/GRAY]
[BOLD_RED]❌ Final Error:[/BOLD_RED] Critical error"
`;

exports[`Visual Formatter Service Core Formatting Behaviors should format the overall progress display 1`] = `
"[BOLD_BLUE]🎯 Icon Generation Progress[/BOLD_BLUE]
[GREEN]████████████[/GREEN][GRAY]░░░░░░░░[/GRAY] [BOLD]60%[/BOLD]
[GRAY]Current Phase: [BLUE]VALIDATION[/BLUE][/GRAY]
[GRAY]Session: test-session-123[/GRAY]
[GRAY]Elapsed: 1m 0s[/GRAY]"
`;
